File: 백준/Gold/1253. 좋다/좋다.py
import sys

# 메인 아이디어: 단순 구현한다면 시간복잡도가 O(N^3)으로 40억을 초과한다.
# 따라서 정렬을 한 후에 두 포인터(초기값 0)를 오른쪽으로 옮겨가며 합이 어떤 수와 같은지 비교한다. 최대 O(N^2)
# tail이 어떤 수보다 같거나 클 수 없다!
inputF = sys.stdin.readline

n = int(inputF().rstrip())
nums = list(map(int, inputF().rstrip().split()))
nums.sort() # 투포인터 사용을 위한 정렬
result = 0

# 각 위치에서 "좋다"인지 확인
for i in range(n):
    left, right = 0, n-1
    while left < right:
        # 포인터가 어떤 수에 걸치면 건너뛴다.
        if left == i:
            left += 1
            continue
        elif right == i:
            right -= 1
            continue

        # 합이 같으면 결과를 ++하고 break
        if nums[left] + nums[right] == nums[i]:
            result += 1
            break

        # 합이 어떤 수보다 작으면 left를 ++
        if nums[left] + nums[right] < nums[i]:
            left += 1
            continue
        # 합이 어떤 수보다 크면 right를 --
        elif nums[left] + nums[right] > nums[i]:
            right -= 1
            continue

print(result)

File: 백준/Gold/1253. 좋다/좋다.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        // 입력 처리
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int[] nums = new int[n];

        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            nums[i] = Integer.parseInt(st.nextToken());
        }

        Arrays.sort(nums);  // 오름차순 정렬
        int result = 0;

        // 각 숫자에 대해 "좋은 수"인지 검사
        for (int i = 0; i < n; i++) {
            int left = 0;
            int right = n - 1;

            while (left < right) {
                // 현재 검사 중인 수(nums[i])는 제외
                if (left == i) {
                    left++;
                    continue;
                }
                if (right == i) {
                    right--;
                    continue;
                }

                int sum = nums[left] + nums[right];

                if (sum == nums[i]) {
                    result++;
                    break;
                } else if (sum < nums[i]) {
                    left++;
                } else {
                    right--;
                }
            }
        }

        System.out.println(result);
    }
}
